File: problems/linked-list/0143-reorder-list.py
"""#143 Reorder List
Split at the middle, reverse the second half, and weave the two halves.
Time: O(n) | Space: O(1)
"""


class Solution:
    def reorderList(self, head):
        if head is None or head.next is None:
            return

        slow = fast = head
        while fast.next is not None and fast.next.next is not None:
            slow = slow.next
            fast = fast.next.next

        second = slow.next
        slow.next = None

        previous = None
        while second is not None:
            following = second.next
            second.next = previous
            previous = second
            second = following

        first, second = head, previous
        while second is not None:
            first_next, second_next = first.next, second.next
            first.next = second
            second.next = first_next
            first, second = first_next, second_next

File: problems/linked-list/0143-reorder-list.js
/**
 * #143 Reorder List
 * Pattern: split at the middle, reverse the second half, then weave both halves.
 * Time: O(n) | Space: O(1)
 *
 * @param {ListNode|null} head
 * @return {void}
 */
function reorderList(head) {
  if (head === null || head.next === null) return

  let slow = head
  let fast = head
  while (fast.next !== null && fast.next.next !== null) {
    slow = slow.next
    fast = fast.next.next
  }

  let second = slow.next
  slow.next = null

  let previous = null
  while (second !== null) {
    const next = second.next
    second.next = previous
    previous = second
    second = next
  }

  let first = head
  second = previous
  while (second !== null) {
    const firstNext = first.next
    const secondNext = second.next
    first.next = second
    second.next = firstNext
    first = firstNext
    second = secondNext
  }
}

module.exports = reorderList
